specs/executionAPISpec.md

```diff
@@ -0,0 +1,530 @@
+* [EFT-7] eth_filterTransaction **MUST** error with code -32000 when `data` or `input` or `to` is not specified.
+* [EFT-8] eth_filterTransaction **MUST** error with code -3200 when `data` or `input` provided caused an EVM error when the `gas` is not specified, otherwise it is ignored.
+## <p id="createAccessList">eth_createAccessList</p>
+* [ECAL-1] eth_createAccessList **MUST** return an [EIP-2930](https://eips.ethereum.org/EIPS/eip-2930) access list biased off the given `transaction` and estimated gas cost when using the access list in the `transaction`.
+* [ECAL-2] eth_createAccessList **MUST** use "latest" when the `defaultBlockParameter` is not specified.
+* [ECAL-3] eth_createAccessList **MUST** use the estimated gas when the `gas` is not specified.
+* [ECAL-4] eth_createAccessList **MUST** error with code -32000 when `gas` is too low.
+* [ECAL-5] eth_createAccessList **MUST** error with code -32000 when the `from` address does not have enough Ether to pay for the transaction.
+## <p id="estimateGas">eth_estimateGas</p>
+* [EEG-1] eth_estimateGas **MUST** return the estimated amount of gas the given `transaction` will take to execute.
+* [EEG-2] eth_estimateGas **MUST** error with code -32000 when the `from` address has insufficient Ether to execute the given `transaction`.
+* [EEG-3] eth_estimateGas **MUST NOT** check if the `from` account has sufficient funds when estimating contract deployment.
+* [EEG-4] eth_estimateGas **MUST** use 0x0000000000000000000000000000000000000000 for `from` when it is not specified.
+* [EEG-5] eth_estimateGas **MUST** error with code -32000 when estimating a contract creation that causes an error within the EVM.
+## <p id="signTransaction">eth_signTransaction</p>
+* [ESNT-1] eth_signTransaction **MUST** return both the signed encoded and signed unencoded format of the `transaction` given.
+* [ESNT-2] eth_signTransaction **MUST NOT** use given transaction `type` for transaction type, **MUST** determine transaction type from given transaction parameters.
+* [ESNT-3] eth_signTransaction **MUST** return encoded legacy transactions in the RLP format.
+* [ESNT-4] eth_signTransaction **MUST** return encoded [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) transactions as the transaction `type` concatenated with the transaction type's defined encoding method.
+* [ESNT-5] eth_signTransaction **MUST** use 0x0000000000000000000000000000000000000000 as default `from` address when `from` is null or not specified.
+* [ESNT-6] eth_signTransaction **MUST** error with code -32000 when the `from` address is locked.
+* [ESNT-7] eth_signTransaction **MUST** error with code -32000 when sending transaction using a `from` address that the client does not have the private key for.
+* [ESNT-8] eth_signTransaction **MUST** error with code -32000 when the `gas` is not specified.
+* [ESNT-9] eth_signTransaction **MUST** use 0x0 for `gas` when parameter is null.
```

Comment:
This results in a useless transaction.  This should either return an error or fill in the gas with a gas estimate, or with a reasonable default.

Reply:
For things like this, I recommend changing the line to read:
```suggestion
* [ESNT-9] eth_signTransaction **SHOULD** return a JSON-RPC error when the `gas` parameter is null.
```
If current clients don't do it, we can't reasonably put it as a `**MUST**` but it is a behavior that all clients **SHOULD** do.

Reply:
The document is written to display what Geth currently does. Though while writing this I falsely assumed that using an empty string was the same as using null. Does this sound better?
```
* [ESNT-8] eth_signTransaction **MUST** error with code -32000 when the `gas` is not specified or null.
* [ESNT-9] eth_signTransaction **MUST** use 0x0 for `gas` when empty.
```
I do agree that it should default to something different though.

Reply:
I think something like this would be better.  Essentially, anytime any of the clients do the wrong thing (like not returning an error on a missing required parameter), we should change "**MUST** do the right thing" to "**SHOULD** do the right thing" so that existing clients aren't out of spec.

Anytime we do **MUST** it should be because:
A. All current clients follow the rule.
B. It is the right behavior.

```
* **MUST** return a JSON-RPC error when the `gas` is not specified or null.  This **SHOULD** be error code -32602.
* **SHOULD** return a JSON-RPC error when the `gas` is an empty string.  This **SHOULD** be error code -32602.
```